modules/dxp/apps/search-experiences/search-experiences-web/src/main/resources/META-INF/resources/sxp_blueprint_admin/js/shared/SelectorModal.js

```diff
@@ -0,0 +1,394 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2000 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+import ClayButton from '@clayui/button';
+import ClayEmptyState from '@clayui/empty-state';
+import {ClayCheckbox} from '@clayui/form';
+import ClayLoadingIndicator from '@clayui/loading-indicator';
+import ClayModal, {useModal} from '@clayui/modal';
+import {ClayPaginationBarWithBasicItems} from '@clayui/pagination-bar';
+import ClayTable from '@clayui/table';
+import {ManagementToolbar} from 'frontend-js-components-web';
+import React, {useContext, useEffect, useState} from 'react';
+
+import {DEFAULT_DELTAS} from '../utils/constants';
+import addParams from '../utils/fetch/add_params';
+import fetchData from '../utils/fetch/fetch_data';
+import sub from '../utils/language/sub';
+import ThemeContext from './ThemeContext';
+
+const SelectorModal = ({
+	deltas,
+	fetchItemsURL,
+	initialSelected,
+	locator,
+	multiple,
+	namespaceParams,
+	observer,
+	onClose,
+	onSubmit,
+	title,
+}) => {
+	const [activePage, setActivePage] = useState(1);
+	const [delta, setDelta] = useState(10);
+	const [error, setError] = useState('');
+	const [loading, setLoading] = useState(false);
+	const [resource, setResource] = useState({items: []});
+	const [selected, setSelected] = useState(initialSelected || []);
+
+	const {namespace} = useContext(ThemeContext);
+
+	useEffect(() => {
+		setLoading(true);
+
+		const paramPrefix = namespaceParams ? namespace : '';
+
+		fetchData(
+			addParams(fetchItemsURL, {
+				[`${paramPrefix}page`]: activePage,
+				[`${paramPrefix}pageSize`]: delta,
+			})
+		)
+			.then((response) => setResource(response))
+			.catch((error) => {
+				setError(error);
+			})
+			.finally(() => {
+				setLoading(false);
+			});
+	}, [activePage, delta, fetchItemsURL, namespace, namespaceParams]);
```

Comment:
We should do this in a separate ticket since there'd be many places to update, but we should start making our fetch requests less bug prone. 💪 

In the [React docs](https://react.dev/learn/you-might-not-need-an-effect#fetching-data), they note how this can encounter a race condition. For example, if you change the page multiple times really quickly, it's possible an older request resolves last (meaning if you started from page 1 and continually clicked next page until you reached page 4, you might still end up on page 2).

The loading state avoid this from happening so that's why it's not a priority at the moment, but it might be a good idea to start using the [Clay Data Provider](https://clayui.com/docs/components/data-provider.html) more or creating a hook that includes an [AbortController](https://developer.mozilla.org/en-US/docs/Web/API/AbortController) to cancel requests when a new one is made. (But maybe the Clay Data Provider can handle aborting requests, but not that I saw in the API).